File: frontend/src/api/contentQueue.ts
import { apiGet, apiGetBlob, apiPatchJson, apiPostJson } from "./client";

export type ContentItem = {
  id: string;
  content_type: string;
  title: string;
  status: string;
  approval_mode: string;
  word_count: number | null;
  generated_at: string | null;
  published_at: string | null;
  target_url: string | null;
  body: string | null;
  notes: string | null;
};

export type ContentQueueResponse = { items: ContentItem[] };

export const fetchContentQueue = (): Promise<ContentQueueResponse> =>
  apiGet<ContentQueueResponse>("/api/v1/content-queue/");

export const updateItemStatus = (id: string, status: string): Promise<ContentItem> =>
  apiPatchJson<ContentItem, { status: string }>(`/api/v1/content-queue/${id}/status`, { status });

export const approveAll = (): Promise<{ updated: number }> =>
  apiPostJson<{ updated: number }, object>("/api/v1/content-queue/approve-all", {});

export const purgeShellQueueItems = (): Promise<{ removed: number }> =>
  apiPostJson<{ removed: number }, object>("/api/v1/content-queue/purge-shell-items", {});

export const generateContent = (): Promise<{
  generated: number;
  items: unknown[];
  error?: string;
  warnings?: string[];
}> => apiPostJson("/api/v1/content-queue/generate", {});

export const generateMonthlyTimeline = (): Promise<{
  generated: number;
  weeks: number;
  items: { week: number; type: string; keyword: string; scheduled_for: string }[];
  error?: string;
  warnings?: string[];
  source?: string;
}> => apiPostJson("/api/v1/content-queue/generate-timeline", {});

export const downloadContentPlanExcel = (): Promise<{ blob: Blob; filename: string }> =>
  apiGetBlob("/api/v1/content-queue/export");

export const publishContentItem = (id: string): Promise<{ ok: boolean; type?: string; error?: string }> =>
  apiPostJson(`/api/v1/content-queue/${id}/publish`, {});

File: backend/app/routes/v1/content_queue.py
"""L4: Content queue — rp_content_queue."""

from uuid import UUID

from fastapi import APIRouter, Depends
from fastapi.responses import Response
from sqlalchemy import text

from app.deps import CurrentClientId, DbSession
from app.services.content_export_service import build_content_queue_xlsx, content_export_filename
from app.schemas.content_queue import (
    ApproveAllResponse,
    ContentQueueItem,
    ContentQueueListResponse,
    StatusUpdateRequest,
)
from app.services.content_generation_service import generate_content_for_client
from app.services.content_queue_service import ContentQueueService
from app.services.content_publish_service import publish_queue_item
from app.services.content_timeline_service import generate_monthly_timeline

router = APIRouter()


def get_svc(session: DbSession) -> ContentQueueService:
    return ContentQueueService(session)


@router.get("/", response_model=ContentQueueListResponse)
async def list_queue(
    client_id: CurrentClientId,
    svc: ContentQueueService = Depends(get_svc),
) -> ContentQueueListResponse:
    return await svc.list_items(client_id)


@router.post("/purge-shell-items")
async def purge_shell_queue_items(
    client_id: CurrentClientId,
    session: DbSession,
) -> dict:
    """Remove title-only rows (no `payload.body`) left from old demo seeds."""
    result = await session.execute(
        text(
            """
            DELETE FROM rp_content_queue
            WHERE client_id = :cid
              AND trim(coalesce(payload->>'body', '')) = ''
            RETURNING id
            """
        ),
        {"cid": str(client_id)},
    )
    n = len(result.fetchall())
    return {"removed": n}


@router.post("/generate")
async def generate_content(
    client_id: CurrentClientId,
    session: DbSession,
) -> dict:
    """Generates landing pages + GBP description via Claude for the logged-in client."""
    return await generate_content_for_client(session, client_id)


@router.post("/generate-timeline")
async def generate_timeline(
    client_id: CurrentClientId,
    session: DbSession,
) -> dict:
    """4-week GBP posts + landing pages from Ahrefs suburb keywords."""
    return await generate_monthly_timeline(session, client_id)


@router.get("/export")
async def export_content_plan(
    client_id: CurrentClientId,
    session: DbSession,
) -> Response:
    """Excel download of pending/approved timeline content + photo URLs (review before approval)."""
    data = await build_content_queue_xlsx(session, client_id)
    filename = content_export_filename()
    return Response(
        content=data,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.post("/{item_id}/publish")
async def publish_content_item(
    item_id: UUID,
    client_id: CurrentClientId,
    session: DbSession,
) -> dict:
    """Publish an approved item immediately (GBP or WordPress)."""
    return await publish_queue_item(session, client_id, str(item_id))


@router.post("/approve-all", response_model=ApproveAllResponse)
async def approve_all(
    client_id: CurrentClientId,
    svc: ContentQueueService = Depends(get_svc),
) -> ApproveAllResponse:
    return await svc.approve_all(client_id)


@router.patch("/{item_id}/status", response_model=ContentQueueItem)
async def update_status(
    item_id: UUID,
    body: StatusUpdateRequest,
    client_id: CurrentClientId,
    svc: ContentQueueService = Depends(get_svc),
) -> ContentQueueItem:
    return await svc.update_status(client_id, item_id, body.status)
